src/main/java/net/snowflake/ingest/streaming/internal/ExternalVolumneManager.java

```diff
@@ -0,0 +1,173 @@
+                  .getClientPrefix();
+    } catch (IngestResponseException | IOException e) {
+      throw new SFException(e, ErrorCode.CLIENT_CONFIGURE_FAILURE, e.getMessage());
+    }
+  }
+
+  /**
+   * Given a fully qualified table name, return the target storage by looking up the table name
+   *
+   * @param fullyQualifiedTableName the target fully qualified table name
+   * @return target storage
+   */
+  @Override
+  public StreamingIngestStorage<T, ExternalVolumeLocation> getStorage(
+      String fullyQualifiedTableName) {
+    // Only one chunk per blob in Iceberg mode.
+    StreamingIngestStorage<T, ExternalVolumeLocation> stage =
+        this.externalVolumeMap.get(fullyQualifiedTableName);
+
+    if (stage == null) {
+      throw new SFException(
+          ErrorCode.INTERNAL_ERROR,
+          String.format("No external volume found for table %s", fullyQualifiedTableName));
+    }
+
+    return stage;
+  }
+
+  /**
+   * Add a storage to the manager by looking up the table name from the open channel response
+   *
+   * @param dbName the database name
+   * @param schemaName the schema name
+   * @param tableName the table name
+   * @param fileLocationInfo response from open channel
+   */
+  @Override
+  public void addStorage(
+      String dbName, String schemaName, String tableName, FileLocationInfo fileLocationInfo) {
+    String fullyQualifiedTableName =
+        Utils.getFullyQualifiedTableName(dbName, schemaName, tableName);
+
+    try {
+      this.externalVolumeMap.put(
+          fullyQualifiedTableName,
+          new StreamingIngestStorage<T, ExternalVolumeLocation>(
+              this,
+              this.clientName,
+              fileLocationInfo,
+              new ExternalVolumeLocation(dbName, schemaName, tableName),
+              DEFAULT_MAX_UPLOAD_RETRIES));
+    } catch (SnowflakeSQLException | IOException err) {
+      throw new SFException(err, ErrorCode.UNABLE_TO_CONNECT_TO_STAGE);
+    }
+  }
+
+  /**
+   * Gets the latest file location info (with a renewed short-lived access token) for the specified
+   * location
+   *
+   * @param location A reference to the target location
+   * @param fileName optional filename for single-file signed URL fetch from server
+   * @return the new location information
+   */
+  @Override
+  public FileLocationInfo getRefreshedLocation(
+      ExternalVolumeLocation location, Optional<String> fileName) {
+    try {
+      ChannelConfigureRequest request =
+          new ChannelConfigureRequest(
+              this.role, location.dbName, location.schemaName, location.tableName);
+      fileName.ifPresent(request::setFileName);
+      ChannelConfigureResponse response = this.snowflakeServiceClient.channelConfigure(request);
+      return response.getStageLocation();
+    } catch (IngestResponseException | IOException e) {
+      throw new SFException(e, ErrorCode.CLIENT_CONFIGURE_FAILURE, e.getMessage());
```

Comment:
please create a channel configure failure message

Reply:
The failure message will be thrown [here](https://github.com/snowflakedb/snowflake-ingest-java/blob/ac7bfc18673d17db16d889fe0b36b42c8121db12/src/main/java/net/snowflake/ingest/streaming/internal/SnowflakeServiceClient.java#L97) if there's no `IOExcetion`. Otherwise it'll just throw error message like other callsites ([ref1](https://github.com/snowflakedb/snowflake-ingest-java/blob/ac7bfc18673d17db16d889fe0b36b42c8121db12/src/main/java/net/snowflake/ingest/streaming/internal/SnowflakeStreamingIngestClientInternal.java#L390), [ref2](https://github.com/snowflakedb/snowflake-ingest-java/blob/ac7bfc18673d17db16d889fe0b36b42c8121db12/src/main/java/net/snowflake/ingest/streaming/internal/SnowflakeStreamingIngestClientInternal.java#L429))